### PR TITLE
docs/adr: ADR for devnet Celestia Core <> Celestia Full node communication

### PR DESCRIPTION
**Pull request by renaynay**
This PR documents an architecture decision made regarding the communication pattern between Celestia Core nodes and Celestia Full nodes for devnet.

### CODE REVIEW

File: docs/adr/adr-002-predevnet-core-to-full-communication.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# ADR for Devnet Celestia Core <> Celestia Node Communication 
+
+## Authors
+
+@renaynay
```
Comment (Wondertan) [inline docs/adr/adr-002-predevnet-core-to-full-communication.md:5]: Kindly ask to add me as a coauthor. Even I didn't write a line in this doc, the core idea of embedding a node and having two modes was mine, as per [this](https://github.com/celestiaorg/celestia-node/pull/48#discussion_r702846034) comment and one of our sync discussions.

File: docs/adr/adr-002-predevnet-core-to-full-communication.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+## Decision
+
+Since the flow of information in devnet is unidirectional, where Core nodes provide block information to Celestia Full nodes, the default behaviour for running a Celestia Full node is to have an embedded Core node process running within the Full node itself. Not only will this ensure that at least some Celestia Full nodes in the network will be communicating with Core nodes, it also makes it easier for end users to spin up a Celestia Full node without having to worry about feeding the Celestia Full node a remote Core endpoint from which it would fetch information. 
+
+It is also important to note that for devnet, it should also be possible to run Celestia Full nodes as `standalone` processes (without a trusted Core node endpoint) as Celestia Full nodes should also be capable of learning of block information on a P2P-level from other Celestia Full nodes.
```
Comment (Wondertan) [inline docs/adr/adr-002-predevnet-core-to-full-communication.md:25]: > (without a trusted Core node endpoint)

without trusted or embedded Core node

File: docs/adr/adr-002-predevnet-core-to-full-communication.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+## Detailed Design
+
+* Celestia Full node should be able to run in a `default` mode where Celestia Full node embeds a Core node process
+* Celestia Full node should also be able to be started with a `--standalone` flag to indicate that the Celestia Full node will be running *without* a Core node process
+* Celestia Full node should be able to take in a `--remote.core` endpoint that would indicate to the Full node that it should *not* embed the Core node process, but rather dial the provided remote Core node endpoint.
```
Comment (Wondertan) [inline docs/adr/adr-002-predevnet-core-to-full-communication.md:31]: The param should be `core.remote` instead. The namespace comes first and I envision other configuration params for `core`, besides remote.

Comment (Wondertan) [reply docs/adr/adr-002-predevnet-core-to-full-communication.md:31]:
```suggestion
* Celestia Full node should be able to take in a `--core.remote` endpoint that would indicate to the Full node that it should *not* embed the Core node process, but rather dial the provided remote Core node endpoint.
```

File: docs/adr/adr-002-predevnet-core-to-full-communication.md (hunk 2 above)
Comment (Wondertan) [reply docs/adr/adr-002-predevnet-core-to-full-communication.md:25]:
```suggestion
It is also important to note that for devnet, it should also be possible to run Celestia Full nodes as `standalone` processes (without a trusted remote or embedded Core node) as Celestia Full nodes should also be capable of learning of block information on a P2P-level from other Celestia Full nodes.
```

File: docs/adr/adr-002-predevnet-core-to-full-communication.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+## Decision
+
+Since the flow of information in devnet is unidirectional, where Core nodes provide block information to Celestia Full nodes, the default behaviour for running a Celestia Full node is to have an embedded Core node process running within the Full node itself. Not only will this ensure that at least some Celestia Full nodes in the network will be communicating with Core nodes, it also makes it easier for end users to spin up a Celestia Full node without having to worry about feeding the Celestia Full node a remote Core endpoint from which it would fetch information. 
+
+It is also important to note that for devnet, it should also be possible to run Celestia Full nodes as `standalone` processes (without a trusted Core node endpoint) as Celestia Full nodes should also be capable of learning of block information on a P2P-level from other Celestia Full nodes.
+
+## Detailed Design
+
+* Celestia Full node should be able to run in a `default` mode where Celestia Full node embeds a Core node process
+* Celestia Full node should also be able to be started with a `--standalone` flag to indicate that the Celestia Full node will be running *without* a Core node process
```
Comment (Wondertan) [inline docs/adr/adr-002-predevnet-core-to-full-communication.md:30]: `standalone` can be interpreted in different meanings. I would call it `core.disable` or any other word for disable, but the idea here is to clearly describe that the given option will run without `core`.

Comment (Wondertan) [reply docs/adr/adr-002-predevnet-core-to-full-communication.md:30]:
```suggestion
* Celestia Full node should also be able to be started with a `--core.disable` flag to indicate that the Celestia Full node will be running *without* a Core node process
```

File: docs/adr/adr-002-predevnet-core-to-full-communication.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+* Celestia Full node should be able to take in a `--remote.core` endpoint that would indicate to the Full node that it should *not* embed the Core node process, but rather dial the provided remote Core node endpoint.
+* Celestia Full nodes that rely on Core node processes (whether embedded or remote) should also communicate with other Celestia Full nodes on a P2P-level, broadcasting new blocks that they've fetched from the Core nodes *and* being able to handle broadcasted block-related messages from other Full nodes on the network.
+
+It is preferable that a devnet-ready Celestia Full node is *agnostic* to the method by which it receives new block information. Therefore, we will abstract the interface related to "fetching blocks" so that in the view of the Celestia Full node, it does not care *how* it is receiving blocks, only that it *is* receiving new blocks. 
+
```
Comment (Wondertan) [inline docs/adr/adr-002-predevnet-core-to-full-communication.md:35]: I would also add some positive consequences, like better UX(user won't have a headache from running/configuring two processes), better debugability, and higher control over Core. The only bad thing I see is a little bit more work.

**Review by Wondertan — CHANGES_REQUESTED** [review]

**Review by liamsi — COMMENTED** [review]
Hlib's suggestions makes sense. Otherwise, this PR also LGTM. Thanks for capturing the decsion.

File: docs/adr/adr-002-predevnet-core-to-full-communication.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+## Consequences of embedding Celestia Core process into Celestia Full node
+
+### Positive
+* Better UX for average devnet users who do not want to deal with spinning up a Celestia Core node and passing the endpoint to the Celestia Full node.
+* Makes it easier to guarantee that there will be *some* Full nodes in the devnet that will be fetching blocks from Celestia Core nodes.
+
+### Negative
+* Eventually this work will be rendered useless as communicating with Celestia Core over RPC is a crutch we decided to use in order to streamline interoperability between Core and Full nodes. All communication beyond devnet will be over the P2P layer.
```
Comment (Wondertan) [inline docs/adr/adr-002-predevnet-core-to-full-communication.md:44]: I would argue embedding is not useless long term, as the communications transition over the P2P layer will be done via disassembling tendermint(core) component by component. And by bringing Core code closer to Node core we are one step closer to disassembling, thus to communication transition. But I am ok not to change anything here, as the ADR is not only about embedding.

**Review by Wondertan — APPROVED** [review]

File: docs/adr/adr-002-predevnet-core-to-full-communication.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
+# ADR for Devnet Celestia Core <> Celestia Node Communication 
+
+## Authors
+
+@renaynay @wondertan
```
Comment (Wondertan) [inline docs/adr/adr-002-predevnet-core-to-full-communication.md:5]:
```suggestion
@renaynay @Wondertan
```

Comment (Wondertan) [reply docs/adr/adr-002-predevnet-core-to-full-communication.md:5]: I saw this after the second review, so this is why they are not aggregated into one

**Review by Wondertan — DISMISSED** [review]

File: docs/adr/adr-002-predevnet-core-to-full-communication.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+It is also important to note that for devnet, it should also be possible to run Celestia Full nodes as `standalone` processes (without a trusted remote or embedded Core node) as Celestia Full nodes should also be capable of learning of block information on a P2P-level from other Celestia Full nodes.
+
+## Detailed Design
+
+* Celestia Full node should be able to run in a `default` mode where Celestia Full node embeds a Core node process
+* Celestia Full node should also be able to be started with a `--core.disable` flag to indicate that the Celestia Full node will be running *without* a Core node process
+* Celestia Full node should be able to take in a `--core.remote` endpoint that would indicate to the Full node that it should *not* embed the Core node process, but rather dial the provided remote Core node endpoint.
+* Celestia Full nodes that rely on Core node processes (whether embedded or remote) should also communicate with other Celestia Full nodes on a P2P-level, broadcasting new blocks that they've fetched from the Core nodes *and* being able to handle broadcasted block-related messages from other Full nodes on the network.
```
Comment (liamsi) [inline docs/adr/adr-002-predevnet-core-to-full-communication.md:32]: > broadcasting new blocks that they've fetched from the Core nodes

Celestia nodes do not necessarily broadcast blocks among each other directly (the embedded core nodes do that on the tendermint p2p level). Celestia nodes broadcast headers among each other and need to be able to request block data from each other on demand.

**Review by liamsi — DISMISSED** [review]
Besides one sentece that could cause some confusion in the future, this LGTM!

File: docs/adr/adr-002-predevnet-core-to-full-communication.md (hunk 8 above)
Comment (renaynay) [reply docs/adr/adr-002-predevnet-core-to-full-communication.md:32]: yeah I meant headers. I'll update it

**Review by liamsi — APPROVED** [review]
🙏🏼

**Review by Wondertan — APPROVED** [review]

**Pull request merged**